### DISCUSSION TITLE
`collect_data_files()` not adding folder to bundle

### DISCUSSION THREAD

**Discussion by Moosems** [pyinstaller/pyinstaller#7550] [Help] [answered]
I know this is almost certainly an issue on my end and not with PyInstaller but when I tried adding a static folder with json files and a png to the spec it never got added to the bundle. 

Beginning of DIP.spec
```spec
from PyInstaller.utils.hooks import collect_data_files

datas = []
datas += collect_data_files("DIP_module.static")
```

File that looks for files:
```python
if getattr(sys, "frozen", False):
    folder = Path(sys._MEIPASS)
else:
    folder = Path(__file__).parent.parent

static_folder = Path(f"{(module_dir := Path(__file__).parent)}/static")
```

Error:
```traceback
Last login: Sat Apr  8 13:09:39 on ttys005
/Users/Moosems/Desktop/DIP-dev/dist/DIP.app/Contents/MacOS/DIP ; exit;
Moosems-MacBook-Pro:~ Moosems$ /Users/Moosems/Desktop/DIP-dev/dist/DIP.app/Contents/MacOS/DIP ; exit;
/Users/Moosems/Desktop/DIP-dev/dist/DIP.app/Contents/MacOS/DIP_module/static/default.json
Traceback (most recent call last):
  File "DIP.py", line 36, in <module>
  File "<frozen importlib._bootstrap>", line 1178, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1149, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
  File "PyInstaller/loader/pyimod02_importers.py", line 352, in exec_module
  File "DIP_module/settings.py", line 15, in <module>
FileNotFoundError: [Errno 2] No such file or directory: '/Users/Moosems/Desktop/DIP-dev/dist/DIP.app/Contents/MacOS/DIP_module/static/default.json'
[14082] Failed to execute script 'DIP' due to unhandled exception: [Errno 2] No such file or directory: '/Users/Moosems/Desktop/DIP-dev/dist/DIP.app/Contents/MacOS/DIP_module/static/default.json'
[14082] Traceback:
Traceback (most recent call last):
  File "DIP.py", line 36, in <module>
  File "<frozen importlib._bootstrap>", line 1178, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1149, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
  File "PyInstaller/loader/pyimod02_importers.py", line 352, in exec_module
  File "DIP_module/settings.py", line 15, in <module>
FileNotFoundError: [Errno 2] No such file or directory: '/Users/Moosems/Desktop/DIP-dev/dist/DIP.app/Contents/MacOS/DIP_module/static/default.json'

logout
Saving session...
...copying shared history...
...saving history...truncating history files...
...completed.

[Process completed]
```

**Comment by rokm** [comment, marked as answer]
If this `DIP_module` is not an installed package (e.g., available in standard site-packages path; but is, for example, a local package located next to your entry-point script), then `collect_data_files` called from .spec file will not be able to resolve it. That's because while PyInstaller does add the entry-point script's parent directory to the search path (unless the directory contains an `__init__.py`, in which case we go up the parent chain until we find the directory without `__init__.py`...), this happens during instantiation of `Analysis` in the .spec file. 

So you either need to add the parent directory of `DIP_module` to `sys.path` at the beginning of the .spec file (i.e., before you call `collect_data_files`) yourself, or you should write a hook instead (because hooks are called after search path is extended by PyInstaller).

**Reply by Moosems** [reply]
Yeah its a module in the same folder that's a relative import.

**Reply by Moosems** [reply]
Can you give examples of the options available? The hook seems less hacky (I don't love adding stuff through sys.path) so I might go with that after figuring out how it works.

**Reply by rokm** [reply]
`sys.path` in .spec:

```python
import sys
from PyInstaller.utils.hooks import collect_data_files

sys.path.append(SPECPATH)  # assuming .spec file is in the same directory  as DIP_module

datas = []
datas += collect_data_files("DIP_module.static")

#... the rest of the spec file
```

The hook way:
```python
# extra-hooks/hook-DIP_module.py
from PyInstaller.utils.hooks import collect_data_files
datas = collect_data_files("DIP_module.static")
```

and then pass the directory that contains extra hooks (here it is assumed that it is called extra-hooks and in same directory as the spec file):
```python
a = Analysis(
   ...
   hookspath=['extra-hooks']
   ...
)

**Reply by Moosems** [reply]
How do I do the hooks solely in the spec file?

**Reply by rokm** [reply]
> How do I do the hooks solely in the spec file?

You don't - hooks are separate .py files.

**Reply by Moosems** [reply]
So I have to list each file in the module in hookspath?

**Reply by rokm** [reply]
No, you give the path to directory in which extra hook files are located. And each hook file must be named `hook-{modulename}.py`.

**Reply by Moosems** [reply]
So:

DIP.spec:
```python
# -*- mode: python ; coding: utf-8 -*-
from PyInstaller.utils.hooks import collect_data_files

datas = [collect_data_files("DIP_module.static"),
        collect_data_files("chlorophyll.schemes")]
...
a = Analysis(
    ["DIP.py"],
    pathex=[],
    binaries=[],
    datas=datas,
    hiddenimports=[],
    hookspath=["DIP_module"],
    hooksconfig={},
    runtime_hooks=[],
    excludes=[],
    win_no_prefer_redirects=False,
    win_private_assemblies=False,
    cipher=block_cipher,
    noarchive=False,
)
...
```

hook-DIP_module.py:
```python
...

```

**Reply by rokm** [reply]
Create an extra directory for your hooks, e.g., called "extra-hooks". (You probably don't want to put the hook file into your `DIP_module` package...)

Create a file in that directory, called `hook-DIP_module.py` with the following contents
```
# extra-hooks/hook-DIP_module.py
from PyInstaller.utils.hooks import collect_data_files
datas = collect_data_files("DIP_module.static")
```

And then add the path to this new directory to the list given via `hookspath`. If the directory is next to your spec file, then the relative path is just the name:
```
a = Analysis(
    ["DIP.py"],
    pathex=[],
    binaries=[],
    datas=[],
    hiddenimports=[],
    hookspath=["extra-hooks"],
    hooksconfig={},
    runtime_hooks=[],
    excludes=[],
    win_no_prefer_redirects=False,
    win_private_assemblies=False,
    cipher=block_cipher,
    noarchive=False,
)
``` 
And you can also create a hook for `chlorophyll` (`extra-hooks/hook-chlorophyll.py`), so you don't have to do any collect_data calls in the spec file.

**Reply by Moosems** [reply]
And keep them in `datas` in `DIP.spec`, yes?

**Reply by rokm** [reply]
No, the`datas` in the spec file should be empty, then. (I've corrected my example)

**Reply by Moosems** [reply]
Worked like a charm. Thank you Rokm!
